xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1840,6 +1800,41 @@ void sendDiscoveryRequest(DiscoveryRequestData request) {
     void sendError(Exception error) {
       requestWriter.onError(error);
     }
+
+    @Override
+    public void onNext(DiscoveryResponse response) {
+      final DiscoveryResponseData responseData = DiscoveryResponseData.fromEnvoyProto(response);
+      if (logger.isLoggable(XdsLogLevel.DEBUG)) {
+        logger.log(XdsLogLevel.DEBUG, "Received {0} response:\n{1}",
+            responseData.getResourceType(), respPrinter.print(response));
+      }
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleDiscoveryResponse(responseData);
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(Status.fromThrowable(t));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(Status.UNAVAILABLE.withDescription("Closed by server"));
+        }
+      });
+    }
```

Comment:
What's the benefit of StreamObserver polymorphism? This is duplicate implementation for both V2 and V3.

Reply:
A cleaner and more elegant implementation.

Yeah, this is duplicated but inevitable as the two `StreamObserver` interfaces are really two separate interfaces (generics are just template). It is duplicated in the existing implementation as well.

Possibly more throughout refactor can eliminate the remaining repeated code. But that will probably make it harder to understand and given such a small amount of repeat it doesn't bother to do so.